File: golang/q600/q668.go
package q600

import "sort"

func findKthNumberPlanA(m int, n int, k int) int {
	if k <= 1 {
		return 1
	}
	if k >= m*n {
		return m * n
	}

	arr := make([]int, 0, m*n)
	for i := 1; i <= m; i++ {
		for j := 1; j <= n; j++ {
			arr = append(arr, i*j)
		}
	}
	sort.Ints(arr)
	return arr[k-1]
}

func findKthNumberPlanC(m int, n int, k int) int {
	c := map[int]int{}
	for i := 1; i <= m; i++ {
		for j := 1; j <= n; j++ {
			c[i*j]++
		}
	}
	arr := make([]int, 0, len(c))
	for k := range c {
		arr = append(arr, k)
	}
	sort.Ints(arr)
	i := k
	for i > 0 {
		i -= c[arr[0]]
		if i <= 0 {
			return arr[0]
		}
		arr = arr[1:]
	}
	return arr[0]
}

func findKthNumberPlanD(m int, n int, k int) int {
	return sort.Search(m*n, func(x int) bool {
		count := x / n * n
		for i := x/n + 1; i <= m; i++ {
			count += x / i
		}
		return count >= k
	})
}

File: hs/src/Q600/Q668.hs
module Q600.Q668 (
    findKthNumberPlanA,
    findKthNumberPlanD
) where

import Data.List(sort)

-- 不做任何检查, 简单暴力方案
findKthNumberPlanA :: Int -> Int -> Int -> Int
findKthNumberPlanA m n k = last . take k $ sort [x*y | x <- [1..m], y<-[1..n]]

findKthNumberPlanD :: Int -> Int -> Int -> Int
findKthNumberPlanD m n k = binsearch 0 (m*n) (\x -> (count x) >= k)
    where
        binsearch l r f =
            let
               h = (l+r) `div` 2
            in
                if l < r then
                    if f h then binsearch l h f
                    else binsearch (h+1) r f
                else l
        count x =
            let
                c = (x `div` n) * n
            in
                c + (foldr (+) 0 [ x `div` a | a <-[x`div`n+1..m]])

-- [10..4] 返回的是空集合, 而 [10,9..4] 返回的是 10 到 4 部分, 很智能啊.
